pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -525,35 +519,46 @@ func (c *Controller) GetProxyServiceInstances(proxy *model.Proxy) ([]*model.Serv
 
 			proxyNamespace = pod.Namespace
 			// 1. find proxy service by label selector, if not any, there may exist headless service
-			// failover to 2
+			// failover to 3
 			svcLister := listerv1.NewServiceLister(c.services.informer.GetIndexer())
 			if services, err := svcLister.GetPodServices(pod); err == nil && len(services) > 0 {
 				for _, svc := range services {
 					out = append(out, c.getProxyServiceInstancesByPod(pod, svc, proxy)...)
 				}
 				return out, nil
 			}
+
 		}
-	}
 
-	// 2. Headless service
-	endpointsForPodInSameNS := make([]*model.ServiceInstance, 0)
-	endpointsForPodInDifferentNS := make([]*model.ServiceInstance, 0)
-	for _, item := range c.endpoints.informer.GetStore().List() {
-		ep := *item.(*v1.Endpoints)
-		endpoints := &endpointsForPodInSameNS
-		if ep.Namespace != proxyNamespace {
-			endpoints = &endpointsForPodInDifferentNS
+		// 2. The pod is not present when this is called
+		// due to eventual consistency issues. However, we have a lot of information about the pod from the proxy
+		// metadata already. Because of this, we can still get most of the information we need.
+		// If we cannot accurately construct ServiceInstances from just the metadata, this will return an error and we can
+		// attempt to read the real pod.
+		instances, err := c.getProxyServiceInstancesFromMetadata(proxy)
+		if err == nil {
+			return instances, nil
```

Comment:
why flip this? the idea was to look at proxy metadata first and then ask k8s as asking k8s is costly

Reply:
We need to get locality from node labels.

It is not that costly compared to proxy meta. I looked carefully at `getProxyServiceInstancesFromMetadata`, it does also check service, which is the most costly procedure i think.